django/db/models/sql/query.py

```diff
@@ -498,6 +498,9 @@ def get_aggregation(self, using, aggregate_exprs):
                             annotation_mask |= expr.get_refs()
                     for aggregate in aggregates.values():
                         annotation_mask |= aggregate.get_refs()
+                    for annotation_alias, annotation in self.annotation_select.items():
+                        if not getattr(annotation, "contains_aggregate", True):
+                            annotation_mask.add(annotation_alias)
```

Comment:
TBH, I'm not convinced :dizzy_face:

Reply:
@felixxm I think you have the right approach but I'd suggest maybe more explicit usage of `get_group_by_cols`

```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index fed5ce57eb..e7a3d78d2d 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -496,11 +496,17 @@ def get_aggregation(self, using, aggregate_exprs):
                     if isinstance(self.group_by, tuple):
                         for expr in self.group_by:
                             annotation_mask |= expr.get_refs()
+                    elif has_existing_aggregation:
+                        # Avoid eliding expressions that might have an
+                        # incidence on the implicit grouping logic.
+                        for (
+                            annotation_alias,
+                            annotation,
+                        ) in self.annotation_select.items():
+                            if annotation.get_group_by_cols():
+                                annotation_mask.add(annotation_alias)
                     for aggregate in aggregates.values():
                         annotation_mask |= aggregate.get_refs()
-                    for annotation_alias, annotation in self.annotation_select.items():
-                        if not getattr(annotation, "contains_aggregate", True):
-                            annotation_mask.add(annotation_alias)
                     inner_query.set_annotation_mask(annotation_mask)

             # Add aggregates to the outer AggregateQuery. This requires making
```

Reply:
@felixxm just to make sure, were you able to reproduce with the project provided by the reporter? I see that your adjusted test case avoid the explicit grouping (`values().annotate(Count)`) but the reported case does use explicit grouping.

My proposed suggestion happens to pass against your test but not the reported use case (explicit grouping). I have to admit that I'm a bit surprised that `values("foo").annotate(cnt=Count(), other=F("test"))` doesn't push `"other"` to `self.group_by` like I would have expected.

Reply:
> @felixxm just to make sure, were you able to reproduce with the project provided by the reporter?

Yes.

> I see that your adjusted test case avoid the explicit grouping (`values().annotate(Count)`) but the reported case does use explicit grouping.
> 
> My proposed suggestion happens to pass against your test but not the reported use case (explicit grouping). I have to admit that I'm a bit surprised that `values("foo").annotate(cnt=Count(), other=F("test"))` doesn't push `"other"` to `self.group_by` like I would have expected.

I added a second regression test with explicit grouping.